text/0006-folio-distributed-vs-centralized-configration.md

```diff
@@ -0,0 +1,93 @@
+
+* Settings stored in Okapi's /_/env APIs.
+* Settings stored in Infrastructure (Kubernetes / Rancher) config maps and secrets.
+* Settings stored in module container environment variables.
+* Settings stored in the stripes front-end (stripes.config.js, etc.)
+
+### mod-configuration will be dropped until Ramsons release
+
+mod-configuration is deprecated due to security problems since March 2022. It shall not be used any more to add new configuration variables. Modules still using mod-configuration have to move to other solutions until the Ramsons release.
+
+### Distributed configuration is preferred
+
+Distributed configuration means, that each module stores its configuration values itself, and offers API endpoints to query and store these values. Distributed configuration in a microservice architecture has some advantages:
+
+* The modules can validate the values according to format and dependencies
+* Modules do not depend on a configuration module, hence a better separation of microservices can be achieved
+* Since all API endpoints have to be documented, a basic documentation of possible configuration variables is mandatory
+* Configuration values can be cached, since no other module can change values.
+* Access to configuration values can effectively controlled by permissions defined in the module.
+* Write-only configuration values are possible, like for credentials. The module can offer other operators than reading values like comparing hashes (possible in central configuration too?)
+* Modules can handle upgrade of configuration variable names or values during module upgrades more flexible
+
+Even when there are also some drawbacks on distributed configuration, it is the preferred way to configure backend modules in FOLIO.
+
+### When to use central configuration
+
+mod-settings solves the security problems of mod-configuration. It is the preferred module if configuration variables shall be stored centrally. It is not recommended to develop specialized modules for other central configuration store.
+
+Centralized configuration can only be used for either:
+
+* Non-sensitive information, that are used by many modules or are completely independent of any module. One example are locale settings.
+* Configurations that are specific to a user.
+
+While these configurations can also be stored decentralized in a module, the developer can decide where these values shall be stored.
+
+### Migration
+
+See Timing Section below.
+
+## Risks and Drawbacks
+
+Migrations from old mod-configuration can fail, and therefore tenant upgrades may fail.
+Distributed configuration requires more developer effort than central configuration.
+
+## Rationale and Alternatives
+
+Both central and distributed configurations were discussed. 
+While pure central configuration offer easier access to the complete configuration of a tenant, this is not desireable sice:
+* it is a antipattern in a microservice architecture 
```

Comment:
is it possible to provide a source for this?

Reply:
Yes, a link to a paper is sufficient. I searched the paper for "configuration" and did not get any hits. It would be good to specify the page/section or pull the specific quote from the paper.

Reply:
I think the topic "shared persistence" is the one, that matches to a central configuration module.

Reply:
> I think the topic "shared persistence" is the one, that matches to a central configuration module

What is meant by shared persistence in this context?

My understanding is that shared persistence typically refers to the use of a [shared database](https://www.enterpriseintegrationpatterns.com/patterns/messaging/SharedDataBaseIntegration.html) as an integration technique between services, rather than a service that is used by other services

Reply:
Yes, the typical example is a shared relational database. But is it better to create a microservice, that proxies the shered database? This microservice itself can be seen as a shared persistence.

Reply:
> is it better to create a microservice, that proxies the shered database? 

I wouldn't describe it as better, more as different. Services offer a different integration boundary and abstraction than a shared database

> This microservice itself can be seen as a shared persistence

It could. I tend to think of these patterns as rather different (for the reason above)

As soon as a service offers more than what raw database access does e.g. by providing different means of access or more behaviour then it not only about the storage

(Aside: I am heavily biased _against_ both shared database integration and a service that is a thin-proxy over a database schema)

Reply:
In the referenced paper the aspect "shared database" and the aspect "data ownership" are handled as one antipattern. If you have a shared database, the data are not owned by the module, but the shared database - this is the same for a service that proxies the shared database.
Julian added the two aspects to the reference to the paper. which i think makes it clear.

Reply:
@fgleixner Ok, does that mean that a service using any data that it does not own, either directly to a database or by accessing another service, is considered an anti-pattern?

Reply:
@marcjohnson-kint Yes, data belonging to one service should not be accessed through another service or a shared database. The service should offer access to its data. This is my understanding.

Reply:
@marcjohnson-kint and @okolawole-ebsco is it OK to resolve the conversation?

Reply:
@fgleixner I disagree with the equivalence between a shared database and shared service (that seems to be the interpretation of the paper) and with some of the associated justification for the centralised model being declined

That is likely out of bounds for this stage of the process, and thus, I'm ok with this conversation being resolved

Reply:
@fgleixner Yes, we can resolve this conversation.